### ISSUE TITLE
beforeSuite hooks is not working as expected

### ISSUE THREAD

**Issue by erinev** [webdriverio/webdriverio#3923]
**Environment (please complete the following information):**
 - **WebdriverIO version:** [5.8.1]
 - **Mode:** [WDIO Testrunner]
 - **If WDIO Testrunner, running sync/async:** [sync]
 - **Node.js version:** [10.15.0]
 - **NPM version:** [6.4.1]
 - **Browser name and version:** [Chrome 74]
 - **Platform name and version:** [Windows 10]
 - **Additional wdio packages used (if applicable):** [@wdio/cli, @wdio/local-runner, @wdio/mocha-framework, @wdio/spec-reporter, @wdio/sync]

**Config of WebdriverIO**
WebdriverIO [config](https://github.com/erinev/wdio-selenium-tests-playground/blob/beforeSuite_hook_bug/wdio.conf.js)

**Describe the bug**
Documentation for **beforeSuite** hook states `Hook that gets executed before the suite starts`.  
But in reality **beforeSuite** hook is executed before each spec file (which duplicates **before** hook which is executed before each spec file).  
So beforeSuite has behavior which is not described in the hook description.

**To Reproduce**
100% reproducible scenario is recreated in this repository [branch](https://github.com/erinev/wdio-selenium-tests-playground/tree/beforeSuite_hook_bug).   
To run the tests follow instructions in [readme](https://github.com/erinev/wdio-selenium-tests-playground/blob/beforeSuite_hook_bug/README.md)

**Expected behavior**
**beforeSuite** hook is executed once before each suite: 
* If i specify `--suite allTests` then beforeSuite hook is called once
* if i specify `--suite allTests --suite customSpec` then beforeSuite is called twice

**Log**
![image](https://user-images.githubusercontent.com/4118417/57146254-93636a00-6dcd-11e9-96a1-04e54be1095d.png)

**Additional context**
-

**Comment by CrispusDH** [comment]
Thanks for you issue. I feel a little bit angry from you :) 
Could you share also test what you executes?

**Comment by erinev** [comment]
sory for my sarcasm.. (i just feeling like i totally don't know how to use webdriver-io, because quite few times after reading docs i have an idea how to implement something but it works in unexpected way and proably its problems with my expectations) i updated the issue. 

It the same situation we are discussing in another issue: https://github.com/webdriverio/webdriverio/issues/3887

To have single login before all tests, we found in the documentation that there is a beforeSuite hook and i instantly thought that all our problems will be solved with it.

because we could run single suite all the time `allTests` and we wouldn't need to include different spec files into single one

**Comment by erinev** [comment]
Additional context on our login before all tests use case:

Our company has composite platform (UI) which means that there is a tens of separate UI applications under one domain. (All UI uses same authentication engine - which is create during login).

So our team owns several UI applications inside platform. Login is not part of our scope and we don't assert anything during login process we need just to perform it.

Login is quite expensive operation because after login you need to switch to auto tests campaign (its like separate bucket of data) if we would perform login before each spec file it will make our tests run slower, more prone to login errors, etc.

So thats why we need to login once and then execute our applications test cases with already created session.

**Comment by mgrybyk** [comment]
Guys, what if we introduce GlobalContext object shared across all the processes?

I think it should be useful and help deal with issues like this one.

**Comment by mgrybyk** [comment]
So you want to introduce ability to keep session across multiple spec files instead of killing and spawning it every time?

**Comment by erinev** [comment]
@mgrybyk we actually already implemented it (with a workarround). We have one spec file (which includes another files with describes and we have only 1 spec file running so we get single login functionality before all tests. we have following structure:
![image](https://user-images.githubusercontent.com/4118417/57174400-f9400800-6e46-11e9-8683-ce5fa45788d5.png)
(only one *.spec.js file) our spec filter looks like this: `specs: [specs/**/${args.userType}.spec.js`],`. 
and code inside single spec file:
```
const fs = require('fs');

const testCasesDirectory = 'test-cases';

describe('Agency user -', () => {
  fs.readdirSync(`${global.rootDir}/specs/${testCasesDirectory}`).forEach((file) => {
    require(`./${testCasesDirectory}/${file}`);
  });
});
```
but we don't like it because all our project will need to have same files include logic (it will be duplicated)

but if it is possible and you guys could change this hook behavior to implied by name - **before suite** (not before each file in suite as it is now), we wouldn't need to use this workarround. Because we would always run single suite (which contains multiple spec files) and we would get same functionality (single login before all specs).

**Comment by erinev** [comment]
and it would be great that both hooks before() and beforeSuite() would have such once option (because it would allow to choose which way you want to run tests: grouped by suites o just spec filter)

**Comment by christian-bromann** [comment]
Why do you want to run all your tests in a single session? That sounds pretty brittle to me and not very effective.

**Comment by christian-bromann** [comment]
> To have single login before all tests, we found in the documentation that there is a beforeSuite hook and i instantly thought that all our problems will be solved with it.

What are you trying to do? Login with a single browser before all tests and run with the same session all other tests in parallel?

You can run an automated test before everything else, which does the login step and shares the cookies or any sort of data that allows you to attach the session to the other browser into a file and use that in the actual tests that run in parallel.

**Comment by erinev** [comment]
we want to replicate real users workflow as close as possible. Real users logins to platform and performs some kind of workflow: Create entity A, assign settings to created entity A, search for entity A in the list, copy entity A to another campaign, etc. Users doesn't logout after each step but it uses the same session to perform full workflow, so in order to replicate thi behavior it requires to share session. 

for me logging in, changing context before each file is much more brittle because before every spec file 2 operations (login/switching context) might fail and tests will fail as well, by having single session you are lowering chance that not your team scope (login/switch context) features will fail and impact actual tests.

for these user workflow tests we use `maxInstance: 1` because in order to do full workflow you can't run in parallel (because real users is doing same - 1 user is doing workflow)

for tests which doesn't test full workflow current approach with multiple logins is sufficient (because in those cases you can run 10 instances and tests will finish in the similiar time even though they will perform redundant logins/context switching)

> You can run an automated test before everything else, which does the login step and shares the cookies or any sort of data that allows you to attach the session to the other browser into a file and use that in the actual tests that run in parallel.
even though for me its seems like workarround:
**could you elaborate how this approach could be implemented with current v5 tooling ?**
Having in mind it would not create some limitations like:
* order of spec files (like some spec should run first) and they would allways would perform login step before all tests (with a possiblity to use `describe.only` or `it.only` features to run only some steps)

**Comment by christian-bromann** [comment]
> for me logging in, changing context before each file is much more brittle

This was not my suggestion. My suggestion was to run a test that simulates the login flow and stores all session information somewhere on the file system. After this test was successful you run another test that will set up the session by using the stored session data and applying it to the new browser session.

With that you have a single login test that, if successful will trigger all other tests to run in parallel.

**Comment by erinev** [comment]
i think current workarround with single spec is cleaner then having somehting like 2 tests which must run before all tests (this approach would introduce need to know that those login and preparation tests must be allways runned, and if we want to run single IT in one test file we would need to not forget to run those 2 preparations tests also, it unneeded headache for new developer)

We will keep using our current workarround with single spec file which includes other describes.
